### ISSUE TITLE
Proper way to manage external dependencies

### ISSUE THREAD

**Issue by RenatoUtsch** [Polymer/polymer#1336]
I am trying to manage external dependencies between two of my projects: both my `poly-nvd3` element and my `poly-cloud` element depend on the d3js library. The problem is, if I reload d3, everything crashes.
Upon searching for the recommended way to manage interdependencies, I was surprised to find that there doesn't exist a way to do this.

The idea that works and that I am currently using just temporarily is to load all scripts before importing the components, and the components just expect that the external library is loaded. Like this:
```javascript
<script src="../d3/d3.min.js" type="text/javascript" charset="utf-8"></script>
<link rel="import" href="../poly-cloud/poly-cloud.html">
<link rel="import" href="../polynvd3.html">
```

Doing this works, but it is not a good way to solve the problem, because this destroys the idea that each component is self-contained. Having this in mind, what is the best solution?

I could think of two ways of a 3rd party library working with web components:

------
- Creating a <library>-import bower package
In my example, a `d3-import` bower package would be created that has d3 as a bower dependency and the following d3.html file:
```javascript
<script src="../d3/d3.min.js" type="text/javascript" charset="utf-8"></script>
```
With this, components could import d3 by using:
```javascript
<link rel="import" href="../d3-import/d3.html">
```
This would take advantage of import deduping, and the library would be imported only once. The problem is that a 2nd bower package would have to be mantained for each 3rdparty library. Also, anyone can create this 2nd package, so problems with who mantains it can arise.

------
- Creating a import.html file inside each library
For example, a `import.html` file could be in the d3 library root with the same contents of the above d3.html file, and then I could import d3 just by using:
```javascript
<link rel="import" href="../d3/import.html">
```
This is interesting because the import.html file now is inside the 3rd party library, what seems easier to mantain.

The problem with this approach of exploiting deduping is that a convention has to be created and enforced from the beginning and followed in all libraries, because if a library has different ways of resolving imports, conflicts can arise between them. If Polymer specified a way to do this as a convention to all elements, this problem would be solved.

Is there a better way to manage external dependencies? Can't the Polymer Project enforce one of these dependency management options? This would solve a lot of future headaches with web components.

**Comment by alefwmm** [comment]
+1

**Comment by justinfagnani** [comment]
Generally we wrap such shared libraries in .html files to get the de-duping behavior. I would be nice to have them in the 3rd party library, so if you find they're open to that, then by all means go for it.

We've also experimented with integrating HTML imports with module loaders. We've created an AMD-compliant module definition library called IMD here: https://github.com/PolymerLabs/imd

IMD doesn't do any loading. Used alone, IMD still require that you HTML import your modules, but it could potentially play nice with other loaders which would pull in dependencies not already loaded via HTML imports. Eventually the goal is to have ES6 and HTML imports work naturally together.
